artifacts.go

```diff
@@ -0,0 +1,169 @@
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/mendersoftware/artifacts/parser"
+	"github.com/mendersoftware/artifacts/reader"
+	"github.com/mendersoftware/artifacts/writer"
+
+	"github.com/urfave/cli"
+)
+
+func writeArtifact(c *cli.Context) error {
+	if len(c.String("device-type")) == 0 || len(c.String("image-id")) == 0 ||
+		len(c.String("update")) == 0 {
+		return errors.New("must provide `device-type`, `image-id` and `update`")
+	}
+
+	he := &parser.HeaderElems{
+		Metadata: []byte(`{"deviceType": "` + c.String("device-type") +
+			`", "imageId": "` + c.String("image-id") + `"}`),
+	}
+
+	ud := parser.UpdateData{
+		P:         &parser.RootfsParser{},
+		DataFiles: []string{c.String("update")},
+		Type:      "rootfs-image",
+		Data:      he,
+	}
+
+	name := "mender.tar.gz"
+	if len(c.String("name")) > 0 {
+		name = c.String("name")
+	}
+
+	aw := awriter.NewWriter("mender", 1)
+	return aw.WriteKnown([]parser.UpdateData{ud}, name)
+}
+
+func readArtifact(c *cli.Context) error {
+	return nil
+}
+
+func validateArtifact(c *cli.Context) error {
+	if c.NArg() == 0 {
+		return errors.New("Nothing specified, nothing validated. \nMaybe you wanted" +
+			"to say 'artifacts validate <pathspec>'?")
+	}
+	_, err := os.Stat(c.Args().First())
+	if err != nil {
+		return errors.New("Pathsec '" + c.Args().First() +
+			"' does not match any files.")
+	}
+
+	f, err := os.Open(c.Args().First())
+	if err != nil {
+		return errors.New("Can not open '" + c.Args().First() + "' file.")
+	}
+	defer f.Close()
+
+	ar := areader.NewReader(f)
+	if ar == nil {
+		return errors.New("Can not read artifact file.")
+	}
+	defer ar.Close()
+
+	p := parser.RootfsParser{
+		W: ioutil.Discard, // don't store update anywhere
+	}
+	ar.Register(&p)
+
+	_, err = ar.Read()
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Artifact file '" + c.Args().First() + "' validated successfully")
+	return nil
+}
+
+func run() error {
+	app := cli.NewApp()
+	app.Name = "artifact"
+	app.Usage = "Mender artifact read/writer"
+	app.UsageText = "artifacts [--version][--help] <command> [<args>]"
+	app.Version = "0.1"
+
+	app.Author = "mender.io"
+	app.Email = "contact@mender.io"
+
+	writeRootfs := cli.Command{
+		Name:   "rootfs-image",
+		Action: writeArtifact,
+	}
+
+	writeRootfs.Flags = []cli.Flag{
+		cli.StringFlag{
+			Name:  "update, u",
+			Usage: "Update `FILE`",
+		},
+		cli.StringFlag{
+			Name:  "device-type, t",
+			Usage: "Type of device supported by the update",
+		},
+		cli.StringFlag{
+			Name:  "image-id, i",
+			Usage: "Yocto id of the update image",
+		},
+		cli.StringFlag{
+			Name:  "name, n",
```

Comment:
I'm wondering if we should replace `--name|-n` with `--output-path|-o`. Output seems more appropriate. Name, on the other hand, is bit confusing. One does not know whether is a path or something else, or where the output will be written to.